### PR TITLE
feat: add option to return in order in client and streamer

### PR DESCRIPTION
**Pull request by JoanFM**
**Goals:**
Try returning in the order of requests from Streamer

### CODE REVIEW

File: docs/fundamentals/client/client.md [hunk 1]
```diff
@@ -760,6 +760,34 @@ None
 
 ````
 
+A Client is used to connect to a Flow who processes Documents in an asynchronous and very distributed way. This means that the order in which the Requests are processed may not be the exact same order as they are sent by the Client.
```
Comment (alexcg1) [inline docs/fundamentals/client/client.md:763]:
```suggestion
A Client connects to a Flow that processes Documents in an asynchronous and very distributed way. This means that the order of the Flow processing the requests may not be the same order as the Client sending the requests.
```
I think a short example would be useful here. Parsing that text was not simple.

Wait, when we say "who processes Documents in an asynchronous and very distributed way" does that refer to client or flow?

Maybe just rewrite this bit. There's ambiguity

File: docs/fundamentals/client/client.md [hunk 2]
```diff
@@ -760,6 +760,34 @@ None
...
+However, you can force the order of the results to be deterministic and the same as they enter the Flow by passing `results_in_order` parameter to {meth}`~jina.clients.mixin.PostMixin.post`.
```
Comment (alexcg1) [inline docs/fundamentals/client/client.md:764]:
```suggestion
However, you can force the order of the results to be deterministic and the same as when they enter the Flow by passing `results_in_order` parameter to {meth}`~jina.clients.mixin.PostMixin.post`.
```

**Review by alexcg1 — CHANGES_REQUESTED** [review]

File: docs/fundamentals/client/client.md (hunk 1 above)
Comment (JoanFM) [reply docs/fundamentals/client/client.md:763]: the Flow processes it in an asynchronous and distributed way

Comment (JoanFM) [reply docs/fundamentals/client/client.md]: There is a test already

**Review by alexcg1 — APPROVED** [review]
LGTM 👍

File: jina/serve/stream/__init__.py [hunk 3]
```diff
@@ -206,15 +219,43 @@ def iterating_task_done(task):
...
+                # if return_in_order is needed, we have to check if the response is to be returned
+                response_request_id = await DataRequest.extract_id(response)
+                if response_request_id == request_ids[0]:
+                    # is the response's turn to be yielded
+                    yield response
+                    request_ids.pop(0)
+                else:
+                    # is not your turn, put to the end
+                    responses_ids[response_request_id] = len(responses_list)
+                    responses_list.append(response)
```
Comment (alaeddine-13) [inline jina/serve/stream/__init__.py:247]: I don't see a need for responses_list here. I think we can directly put the response in the responses_ids map (actually let'ts map from ids to response objects and rename the variable)

File: jina/serve/stream/__init__.py [hunk 4]
```diff
@@ -206,15 +219,43 @@ def iterating_task_done(task):
...
+
+                stop_yielding = False
+                while not stop_yielding and len(request_ids) > 0:
+                    next_request_id = request_ids[0]
+                    if next_request_id in responses_ids:
+                        response_index = responses_ids[next_request_id]
+                        del responses_ids[next_request_id]
+                        request_ids.pop(0)
+                        yield responses_list[response_index]
```
Comment (alaeddine-13) [inline jina/serve/stream/__init__.py:256]: assuming we no more need responses_list, we can change the code like so:
```suggestion
                        request_ids.pop(0)
                        yield responses_ids.pop(next_request_id)
```

File: jina/serve/stream/__init__.py [hunk 5]
```diff
@@ -206,15 +219,43 @@ def iterating_task_done(task):
...
+                    request_ids.pop(0)
+                else:
+                    # is not your turn, put to the end
+                    responses_ids[response_request_id] = len(responses_list)
+                    responses_list.append(response)
+
+                stop_yielding = False
+                while not stop_yielding and len(request_ids) > 0:
+                    next_request_id = request_ids[0]
+                    if next_request_id in responses_ids:
```
Comment (alaeddine-13) [inline jina/serve/stream/__init__.py:252]: what happens if the next_request id is not received ? we don't yield anymore ?

File: tests/integration/test_return_order/test_return_order.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+from jina import Flow, Executor, requests, Document, DocumentArray, Client
+import random
+import time
+
+import pytest
+
+
+@pytest.mark.parametrize('protocol', ['grpc'])
+def test_return_order_in_client(protocol):
+    class ExecutorRandomSleepExecutor(Executor):
```
Comment (alaeddine-13) [inline tests/integration/test_return_order/test_return_order.py:10]:
```suggestion
    class RandomSleepExecutor(Executor):
```

File: tests/integration/test_return_order/test_return_order.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+        @requests
+        def foo(self, *args, **kwargs):
+            rand_sleep = random.uniform(0.1, 1.3)
+            time.sleep(rand_sleep)
+
+    f = Flow(protocol=protocol).add(uses=ExecutorRandomSleepExecutor, replicas=2)
```
Comment (alaeddine-13) [inline tests/integration/test_return_order/test_return_order.py:17]:
```suggestion
    f = Flow(protocol=protocol).add(uses=RandomSleepExecutor, replicas=2)
```

File: jina/serve/streamer.py [hunk 8]
```diff
@@ -18,6 +18,25 @@
...
+    return TopologyGraph(
+        graph_representation=graph_description,
+        graph_conditions=graph_conditions,
+        deployments_metadata=deployments_metadata,
+        deployments_no_reduce=deployments_no_reduce,
+        timeout_send=timeout_send,
+        retries=retries,
+    )
+
+
```
Comment (JohannesMessner) [inline jina/serve/streamer.py:39]: I don't understand the purpose of this, it seems like it is the same as just calling the constructor directly?

File: jina/serve/stream/__init__.py [hunk 9]
```diff
@@ -88,14 +94,19 @@ async def stream(
...
     ) -> AsyncIterator:
         """Implements request and response handling without prefetching
         :param request_iterator: requests iterator from Client
+        :param results_in_order: return the results in the same order as the request_iterator
         :yield: responses
         """
         result_queue = asyncio.Queue()
+        request_ids = []
+        responses_list = []
+        responses_ids = {}  # map from id to index in responses_list
```
Comment (JohannesMessner) [inline jina/serve/stream/__init__.py:109]: I would rename to `response_id_to_index` or something like that

File: docs/fundamentals/client/client.md [hunk 10]
```diff
@@ -760,6 +760,34 @@ None
 
 ````
 
+A Client connects to a Flow that processes Documents in an asynchronous and very distributed way. This means that the order of the Flow processing the requests may not be the same order as the Client sending the requests.
```
Comment (JohannesMessner) [inline docs/fundamentals/client/client.md:763]: Is there a way to pass this parameter from a 3rd party client?

Comment (JoanFM) [reply docs/fundamentals/client/client.md:763]: No, this is gonna be ready in the `streamer` for the CustomGateway. We can add it later I would say

Comment (JoanFM) [reply docs/fundamentals/client/client.md:763]: Mostly external clients do not have this `streaming` capability so easily implemented, so I think is not so so important

**Review by alexcg1 — APPROVED** [review]
LGTM :+1:

**Pull request merged**